sklearn/utils/_array_api.py

```diff
@@ -278,33 +372,37 @@ def _asarray_with_order(array, dtype=None, order=None, copy=None, xp=None):
     """
     if xp is None:
         xp, _ = get_namespace(array)
-    if xp.__name__ in {"numpy", "numpy.array_api"}:
+    if _is_numpy_namespace(xp):
         # Use NumPy API to support order
-        array = numpy.asarray(array, order=order, dtype=dtype)
-        return xp.asarray(array, copy=copy)
+        if copy is True:
+            array = numpy.array(array, order=order, dtype=dtype)
+        else:
+            array = numpy.asarray(array, order=order, dtype=dtype)
+        return xp.asarray(array)
```

Comment:
Why can't this be the following (why the extra `xp.asarray()` at the end?):
```suggestion
            return numpy.array(array, order=order, dtype=dtype)
        else:
            return numpy.asarray(array, order=order, dtype=dtype)
```
worth adding a comment to explain to people from the future?

Reply:
That extra `xp.asarray` is required because the input array can be a `numpy.array_api` array. In that case, we convert the `numpy.ndarray` back to the input array's namespace.

I added a comment in [`878da33` (#25956)](https://github.com/scikit-learn/scikit-learn/pull/25956/commits/878da33c1676a50a55d3a26fe3b204cd526a6966).

Reply:
Ah yes. I keep forgetting about `numpy.array_api`.